### ISSUE TITLE
Pythagorean Triplets - How?

### ISSUE THREAD

**Issue by mohmad-null** [exercism/problem-specifications#1902]
https://github.com/exercism/problem-specifications/blob/main/exercises/pythagorean-triplet/description.md

Does not explain how to actually generate Pythagorean Triplets. Being maths naive and before I had properly considered the problem I figured it was just a "iterate through all possible combinations". This is unfeasible when there are 27 trillion possible combinations to iterate through (for the 30,000 test).

I'm sorry but this description of the problem is woefully lacking for anyone who doesn't have some advanced maths. From it I can certainly *test* whether something is a Pythagorean Triplet, but there's no hint as to how you're meant to do so.

**Comment by siebenschlaefer** [comment]
Did you try to research the problem, e.g. by asking your favorite search engine for "generate pythagorean triples"?

**Comment by mohmad-null** [comment]
> Did you try to research the problem, e.g. by asking your favorite search engine for "generate pythagorean triples"?

Thank you for the suggestion, but this circumvents the point: It shouldn't be necessary! This is a website for learning programming, not weird maths formulae that I'll never need to use again.

(And that's ignoring that many of the responses are written using notation and language that requires some proficiency with maths - looking at you [wikipedia](https://en.wikipedia.org/wiki/Formulas_for_generating_Pythagorean_triples))

**Comment by siebenschlaefer** [comment]
Quoting from the exercise "difference-of-squares":
> You are not expected to discover an efficient solution to this yourself from first principles; research is allowed, indeed, encouraged. Finding the best algorithm for the problem is a key skill in software engineering.

I think the same applies to a lot of other exercises, too, and it is also true in practice.

My *personal* opinion: If the domain-specific knowledge can be acquired easily I don't see it as a problem. For this exercise googling "generating pythagorean triples" will lead you to the dedicated page on Wikipedia. I think this exercise is fine as it is.

**Comment by SaschaMann** [comment]
> Thank you for the suggestion, but this circumvents the point: It shouldn't be necessary! This is a website for learning programming, not weird maths formulae that I'll never need to use again.

Practice exercises sometimes require additional knowledge. Some require knowledge on how to write a parser, knowledge about strings, floating point arithmetics etc. All of those are topics that other people "will never need to use again". It heavily depends on your background and what you want to learn programming for. Someone who uses Exercism to improve their programming skills so they can write better code for their research might not need to do any of the string exercises, yet there are dozens of them.

You don't have to do any of them so if you don't want to solve more mathsy exercises, you can simply skip them. But giving away the solution will take away exercises from people who do enjoy like this.

---

That said, finding a solution that runs fast enough is possible with only 7th grade maths and no knowledge of advanced notation, see the 2nd example here for example: https://github.com/exercism/website-copy/blob/main/tracks/julia/exercises/pythagorean-triplet/mentoring.md (the code is in Julia but the explanation should be understandable without it)

**Comment by mohmad-null** [comment]
> "generating pythagorean triples" will lead you to the dedicated page on Wikipedia. 

I linked to it in my last message as an example of how opaque results can be (Wikipedia is usually the worst and this example doesn't fail to fail). Evidently you are able to parse that but I can assure you, it's a wall of unhelpful gibberish to me and I would be more than happy to bet, to most other people on the planet too.


> https://github.com/exercism/website-copy/blob/main/tracks/julia/exercises/pythagorean-triplet/mentoring.md

The __Cubic time solution__ is the one I would have come up with myself, except there's an implicit test to ensure that you don't actually use that solution (at least with  Python; `n=30,000`). So the tests have arbitrarily decided that the user *must* have a certain level of maths proficiency; the naive solution is unacceptable.

> Practice exercises sometimes require additional knowledge.

Indeed, but where is that limit? I put it to you that they should all be self-contained as much as possible.

I suspect the reason that maths problems seem to fall through the cross-domain "acceptability" gap is that many folks who code also happen to be highly proficient at it. You need to remember that most people don't share that capability.

**Comment by junedev** [comment]
@mohmad-null Its a bit difficult to discuss in two issues at once. Since the discussion started here, could you integrate the other examples into this issue please and close the other one (you can also adjust the title here). Then we can continue the discussion here. Also please remember the [Code of Conduct](https://exercism.org/docs/using/legal/code-of-conduct) and make sure to be kind and polite. (Applies to everyone of course.)

Now to the topic. I think @mohmad-null raises some valid points.
* a) Besides some sentences in some exercises (which the student might never see), see quote from @siebenschlaefer, there is no good explanation that a student is supposed to research stuff in practice exercises. If we expect this, there are two possibility how this could be addressed:
  * Have a general, easy to find and clear explanation of what practice exercises are and what is expected of a student to solve them somewhere on the website. I looked for this recently and couldn't find anything that really reflects the current state of practice exercises.
  * Add a sentence similar to the quote above to all the exercises where they apply.
* b) I think it is very true that while you can find information on Wikipedia etc. it can require quite some understanding of math to make sense of the formulas there. So we should check carefully whether googling really gives a reasonable way of solving the exercise without great math skills. If not, then a specific link should be provided or the exercise should clearly be labelled as "mathy" somehow.

**Comment by SaschaMann** [comment]
> Indeed, but where is that limit? I put it to you that they should all be self-contained as much as possible.

I think we need to differentiate between explaining the solution and the terms mentioned in the problem description.

There are a few things in this exercise's description that aren't explained in it and not linked anywhere:

> A Pythagorean triplet is a **set** of three **natural numbers**, {a, b, c}, for which, [...] Given an input **integer** N, find all Pythagorean triplets for which a + b + c = N.

Adding explanations or links to explanations to those concepts seems reasonable to me. However, actually giving away the _solution_ in the description makes this exercise uninteresting to anyone who likes to figure out things on their own or with some research and turns this into an algorithm-implementation exercise for _everyone_. I might have understood you but as far as I understand, you're suggesting the description should contain an explanation for a better-than-cubic-time solution, which imo falls under the latter.

There's always the option to submit the slow solution and request mentoring, too.

---

> (Wikipedia is usually the worst and this example doesn't fail to fail)

I agree. I think links to Wikipedia should be avoided in for explanations of prerequisite concepts, there are usually sites that explain it without as much lingo.

**Comment by mohmad-null** [comment]
Searching for the results has a number of problems:

1) It's adding a further barrier to the problem, impeding the problem solving process.
2) The barrier is unnecessary. Unless you're trying to teach people how to solve problems in a specific domain, there's no reason for a problem to be in domain specific.
2) It requires a certain education & education level to be able to integrate whatever the technical information is into your mental models
3) Search results are susceptible to the filter bubble. What you find isn't what I'll find, even for the exact same term.
4) It ignores global inequalities in education access
5) It ignores global inequalities in computer/internet access (searching requires more time/bandwidth which can often be highly constrained).
6) It means the exercises can't be done offline.

For all of these reasons I'd again like to suggest all problems should be self-contained. There's absolutely no reason they shouldn't be.

"But it's n-grade maths" exemplifies it well, in particular:
a) Anyone who hasn't reached that grade can't participate.
b) It assumes all countries teach at the same levels
c) There is a *vast* difference between maths proficiency levels, both within countries and globally. I point you to Figure 3 here (8th grade maths globally) - https://nces.ed.gov/programs/coe/indicator/cnt

>  However, actually giving away the solution in the description makes this exercise uninteresting to anyone who likes to figure out things on their own

If a person likes solving maths problems, more power to them. However, I'd like to suggest there are probably a lot of websites out there filled with tons of interesting maths problems for them to solve. Same for quite other domains, albeit to a lesser degree than maths.
Exercism on the other hand is a generalist site designed to teach people *programming*.

> I understand, you're suggesting the description should contain an explanation for a better-than-cubic-time solution .... There's always the option to submit the slow solution and request mentoring, too.

For Pythagorean Triplets, yes. The alternative is to remove the tests that make the maths-naive solution unworkable. It's impossible to submit the naive solution: it won't complete [because n=30,000 in one of the Python Tests.](https://github.com/exercism/python/blob/6908da217fac3d27b8ef749f1e3a10ae7dc62829/exercises/practice/pythagorean-triplet/pythagorean_triplet_test.py#L47)

**Comment by SaschaMann** [comment]
> If a person likes solving maths problems, more power to them. However, I'd like to suggest there are probably a lot of websites out there filled with tons of interesting maths problems for them to solve. Same for quite other domains, albeit to a lesser degree than maths.
> Exercism on the other hand is a generalist site designed to teach people _programming_.

Programming doesn't stand on its own. It's always applied to some field or subject. There's no universal background or field one can assume. I'm repeating myself but if you say that if this is an issue with maths exercises, the same must be applied to all exercises. You can take a programming course and strings will never even be mentioned, yet it's still programming and you learn how to program in it. We don't give away the solution to `bob` or other string exercises in the description. Nor do we explain what strings are.

Nobody is forcing you to solve all practice exercises in a track. You can skip them, only do the ones that interest you etc. Why should we _exclude_ those people from Exercism and send them to other sites? The Python track has 129 exercises, the vast majority of which don't require any kind of maths or arithmetics at all. The barrier and exclusion you are outlining only applies if this was somehow a gatekeeper exercise but it isn't. For exercises that do block progression (Learning/Concept exercises), there are far stricter rules on prerequisite/background knowledge.

As a side note, it's not a site designed to teach people programming. It's a site designed to teach people fluency in a programming language. It's pedantic but actually makes a difference because it allows us to assume certain knowledge. Having programming knowledge is required already, it's not a site for learning from scratch. For example, you will find that _concept_ exercises that introduce loops in a track generally won't cover what a loop actually is, they will cover how a loop works in that particular language.

The problem that a naive brute-force approach doesn't solve a particular programming problem isn't restricted to mathsy problems either. I don't know if there is one in problem-specs but it's certainly possible to design exercises that aren't mathsy at all that will have this problem.

> It's impossible to submit the naive solution: it won't complete because n=30,000 in one of the Python Tests.

You can submit solutions that don't pass the tests via the CLI to request mentoring. You can also look up community solutions without having solved the exercise by [directly going to the link](https://exercism.org/tracks/python/exercises/pythagorean-triplet/solutions). I don't agree on these design choice to hide these options but that's been discussed in other issues.

**Comment by siebenschlaefer** [comment]
@moshegood How would this exercise look if it were self-contained? Are you thinking of something like this:

> A Pythagorean triplet is a set of three natural numbers, {a, b, c}, for which,
> ```
> a² + b² = c²
> ```
> and such that,
> ```
> a < b < c
> ```
> For example,
> ```
> 3² + 4² = 9 + 16 = 25 = 5².
> ```
> Given an input integer N, find all Pythagorean triplets for which a + b + c = N.
> 
> For example, with N = 1000, there is exactly one Pythagorean triplet for which a + b + c = 1000: {200, 375, 425}.
>
> One possible algorithm that generates these Pythagorean triplets works like this
> ```
> let a ← 3
> while a < n ÷ 3
> do
>     let b ← (n² - 2n × a) / (2 × (n - a))
>     if a < b and ⌊b⌋ = b
>     then
>         c ← n - a - b
>         append(result, (a, b, c))
>     endif
>     a ← a + 1
> done
> ```

---

And in general: Do you have a solution for this problem that you receive?

- Do you think problems like this should be removed from Exercism?  
- Do you think problems like this should be put on some kind of side-track? 
- Do you think the instructions should provide a link to some webpage where possible algorithms are explained?
- Do you think the instructions themselves should present a possible algorithm so that the student can focus on implementing it?

**Comment by mohmad-null** [comment]
> Programming doesn't stand on its own. It's always applied to some field or subject.

Indeed yes, but then the question again becomes: Why are there so many maths-centric problems?
Where are the problems for NLP, GIS, data analysis, visualisation (ok, that'd be hard to test), networking, Geology, Paleobotany, taxonomy, music theory (1), material science, philately,... ad infinitum?

If you head down this road then you open the door to *all* programming problems. Issue there is, it becomes harder for everyone who doesn't know about the 99% of fields that aren't theirs to find exercises that are suitable for them. How are you going to find suitable problems when there are 200, 400, 700, 1000? Intermingled across innumerable domains? I already struggle with Python.


> Nobody is forcing you to solve all practice exercises in a track.

There are few problems with this reasoning, largely running around the fact it doesn't mesh with how many humans actually work.

* Some people do have a compunction forcing them to complete all of them. Especially given Exercism is heavily gamified.
* I now have some of these permanently "in progress" and no way to remove them; I expect other users do too.
* As above, the more domain-specific ones there are, the harder it becomes for all the users who don't have/want that knowledge to find things they *can* do. It's *already* hard for me to do this on Python (hence this ticket).
* And the more non-domain problems you solve, the more domain ones are left making the remainding non-domain problems even harder to find.
* As in my first post, the frustration from these can increase stress, frustration, and facilitate Imposter Syndrome.


---------

As to @siebenschlaefer questions.

My answer remains:
* Entirely self-contained exercises
* No domain specific knowledge required

Anything else and Exercism is failing at its stated goal of inclusion of people "regardless of their background or motivations".

If you want domain specific problems, cool, but put them somewhere slightly different on the site (another tab?) to make it clear there's a distinction. This will address the problems raised above. I know I'd be happy to solve problems in certain domains (not maths! ;-) ), and a section of the site filled with them would be great, but not intermingled with everything else.

**Comment by SaschaMann** [comment]
> I now have some of these permanently "in progress" and no way to remove them; I expect other users do too.

I think you should raise this as a suggestion on exercism/exercism. Having some kind of hide button might be nice regardless of the outcome of this discussion.

**Comment by BethanyG** [comment]
> _For all of these reasons I'd again like to suggest all problems should be self-contained. There's absolutely no reason they shouldn't be._

@mohmad-null   -- Could you provide some examples of practice exercises (_not the concept exercises_) that currently meet this criteria?  `Self-contained` could be interpreted as only practicing a given programming language _syntax_, as opposed to gaining practice in language _use_ by solving a problem.  

While I don't think that's your intent, it would be really helpful to me to see what you consider exemplar exercises.  As I think about the 117-odd practice exercises on the Python track, very few of them feel self-contained to me.

I'd also love to see any programming problems from other sites or resources you can point to.  As we create more concept and practice exercises, it would be good to have examples.

**Comment by iHiD** [comment]
@mohmad-null Thanks for raising your concerns! We don't learn how others are experiencing our exercises unless people tell us, so it's essential for us to get feedback like this. So thanks for taking the time 🙂 

And thanks to everyone else for digging into the questions raised.

I think there's a few different threads here, but probably two main ones:
1. Maths stuff is painful for people without much maths knowledge (such as myself).
2. Exercises that require other (non-maths) domain knowledge need to have that knowledge clearly explained.

For (1) I agree entirely and have wanted to "fix" this for a long time. The challenge is that maths is very common for some programmers and very uncommon for other programmers. In a language like Python, which is used for data science (maths is common) and all sorts of general purpose programming (maths generally uncommon), these exercises are brilliant for some people and a nightmare for others. So for @SaschaMann, who's into maths, he's like "this is obvious" to a child. I, however (despite having a CS degree) now have no memory of what a "natural number" is, find that exercise immediately alienating. We've previously discussed tagging maths-based exercises with a "maths" flag, which shows up as a sort of "If you like maths, you're gonna love this. If not, you might want to skip it" sort of heading (or something) and maybe the ability to disable those exercises in your settings. I think that potentially achieves the best of both worlds. I don't want to remove maths exercises as I think they are useful in many languages where those languages are commonly learnt because people are using it for math-y things, but I agree that they're really alienating to lots of people.

For (2) I don't think there are **many** other exercises that require domain-knowledge that isn't generally well explained. I've edited ones in the past where I found it to be too hard to parse (e.g. [Hamming](https://github.com/exercism/problem-specifications/pull/1360)) and would love to find time to do this with a wider set. But I do also agree with the idea that it's nice to have exercises on topics that you might need to go and learn something about, as long as they sort of feel like more "projects" or "bonus" exercises, than something fundamental. Again, it feels like the key is making this clear to students and ensuring that the instructions link to places that explain things easily from first principles (e.g. not Wikipedia). 

On a more general note, I agree with what @BethanyG nicely surmised, that it's good to practice problem solving in a programming language, as well as practicing syntax. ​It's probably important for tracks to consider how they're structured to ensure that we learn syntax then get to practice things, and maybe the ordering of practice exercises (or us adding further sub-categorisations) might help with that. I know re Python that Bethany puts a lot of thought into that structuring, so maybe it's more of a website-level thing to fix.

---

You ask why there are so many maths ones. Because maths exercises are easier to write (as they're normally existing problems like Pythag triplets that are commonly used) and because Exercism was seeded with lots of exercises like these. Newer exercises (e.g. the Concept Exercises) are very rarely mathsy. In fact, I'd suggest that nearly all "original" exercises on Exercism are not-mathsy, but ones that were taken from elsewhere early on are mathsy.

---

@mohmad-null Finally I think your points in the first half of [this post](https://github.com/exercism/problem-specifications/issues/1902#issuecomment-1001244099) are a fantastic reminder to us of which this subject is important. It's something everyone here cares about, but it's hard for us to sometimes put ourselves in the mindsets of others with extremely different experiences, to understand what is easy/hard/exciting/offputting for others. So thank you for raising those 🙂

**Comment by mohmad-null** [comment]
> Could you provide some examples of practice exercises (not the concept exercises) that currently meet this criteria? Self-contained could be interpreted as only practicing a given programming language syntax, as opposed to gaining practice in language use by solving a problem.

Sure. I'd suggest to me that all of the below are fairly well self-contained. At least by my definition.

##### Basic Ones
* Isogram
* Word Count
* Raindrops
* Matrix
* Pangram
* Series
etc etc

##### Domain Specific ones
Some are already both self-contained and domain specific, although some of them don't have the best explanations/descriptions.
* Hamming (Genetics)
* ISBN Verifier
* Luhn
* Collatz Conjecture (Maths)
* Triangle (Geometry)
* Prime Factors (Maths)
* Queen Attack (Chess)
* Leap (Calendars)
* RNA Transcription
* Yacht (Game)
* Resistor Color (Electronics)
Etc

##### Borderline
* Grains (It's self-contained but it requires understanding of exponents, whereas the others are just +-*/ )
* Markdown (Parser) - strictly speaking there's no need to know markdown before doing this because it's already written and you can figure it out from the code.

So as you can see, they're definitely possible.

Of course, that's just my (subjective) list. What'd be really interesting is stats for how many people start / complete any given exercise. A quick search suggests there are none :-( .
